gcs_light_client.mediawiki

```diff
@@ -0,0 +1,1056 @@
+    // If we reach this point, then the item isn't in the set.
+    false
+</pre>
+
+===== Querying Against a Set of Items =====
+
+For most applications, the common case will be attempting to match a ''list'' of
+items to the filter. In this case, we can perform a "zip" search against two
+sorted lists: the step-by-step decompressed values of the set, and the list of
+items we'd like to query.
+
+The following routine will evaluate to ''true'' if ''any'' of the items in a
+target set are ''maybe'' within the original set of items (pre encoding):
+<pre>
+gcs_match_any(key: [16]byte, compressed_set: []byte, targets [][]byte, 
+              fp, N: int) -> bool:
+
+    stream = new_bit_stream(compressed_set)
+
+    // Once again, we'll map our set of target values into the domain our
+    // encoding, sorting as a last step so we can zip through the values.
+    let items = []
+    for t in target:
+        let item = (siphash(key, t) * (N * (1 << fp))) >> 64
+        items.append(item)
+    items.sort()
+
+    // Set up a set of accumulator values that we'll use to zip down the two
+    // filters.
+    let last_set_val, last_target_val = 0, 0 
+    last_target_val = items[0]
+    let = 1
+
+    // We'll keep running until one of the values matches each other. If this
+    // happens, then we have a match!
+    while last_set_val != last_target_val:
+        // Perform a pattern match to decide which filter we'll need to
+        // advance.
+        match:
+            case last_set_val > last_target_val:
+                // If we still have items let, advance the pointer by one.
+                if i < len(items):
+                    last_target_val = items[i]
+                    i++
+
+                // Otherwise, we've ran our items in our target set, which
+                // means nothing matched.
+                false
+
+            case last_target_val > last_set_val:
+                // In this case, we'll advance the filter we're querying
+                // against. This entails decompressing the next element in the
+                // set.
+                let decoded_value = golomb_decode(stream, fp)
+
+                // Accumulate the decoded delta value to the current value in
+                // order to retrieve the current set item.
+                last_set_val += decoded_value
+
+    // If we reach this point, the two items in the set matched!
+    true
+</pre>
+
+
+=== Peer to Peer Network Extensions ===
+
+With the procedures to construct, compress, and query the sets explained, we'll
+now turn to the modifications to Bitcoin's p2p protocol required to support
+this new operating mode.
+
+
+==== Peer to Peer Service Bit ====
+
+To start, we reserve a currently unutilized service bit. This is required as
+light clients SHOULD ''preferentially'' peer to full-nodes that support the
+features outlined in this BIP.
+
+The 6th service bit will now be dedicated to signaling support for the
+features described within this BIP: 
```

Comment:
I agree. It would be ideal if pruned nodes could set this and still be usable by light wallets (I might even suggest that clients SHOULD support using pruned nodes... many users want pruned nodes, and their own node should be the expected server)